### DISCUSSION TITLE
[tvOS] How to config for react-native-tvOS?

### DISCUSSION THREAD

**Discussion by nhannfq** [invertase/react-native-firebase#5722] [Q&A]
I'm trying build a tvOS application for the Apple TV, react-native-firebase doesn't seem to have support forfor tvOS, does it?

**Comment by mikehardy** [comment]
I would be shocked if it worked. It's on my wishlist to provide that but I haven't had time.
I believe the work to do, if you've got the time, is mostly going in and adding the appropriate ifdefs/ifndefs etc that toggle out apple API symbols that are (or are not) defined for tvOS, along with appropriate markup in associated `index.d.ts` files (that are processed into our API documentation) noting where things are not supported on tvOS, perhaps with a main docs page in the core/app module about tvOS specifically that links to the upstream firestore-ios-sdk package documentation on what modules work for tvOS and what don't, and what the differences are from the iOS platform

**Reply by hotrush** [reply]
Hope you'll find some time 🥲

**Reply by mikehardy** [reply]
Hey @hotrush I'm hoping to branch out and do more multi-platform stuff this year (web is first priority - https://github.com/invertase/react-native-firebase-authentication-example, but then the other Apple platforms seem like a natural fit here - macos + tvOS). I have a little bit of experience with react-native-macos now but I haven't done any react-native-tvos work.

Can you tell me what your basic project init expectations are? Do the current instructions from react-native-tvos work for you, to init a project? https://github.com/react-native-tvos/react-native-tvos#react-native-tvos

Is it really true that you cannot do tvos and core at the same time, implying that if I init a tvos project then that native directory won't work for ios builds?

**Reply by hotrush** [reply]
I used react-native-tvos and was able to build and run it on emulator, seems to be fine on first look. Also I built it and ran for ios, so that's possible, but not sure that makes a lot of sense. 

You can check this starter to have a quick look https://github.com/react-native-tvos/rnn-starter-tv

**Reply by nriccar** [reply]
Hello to all! Is this still not possible?

**Reply by mikehardy** [reply]
This is probably still not possible. No attempt has been made yet to go through our ios code and `#ifdef TV_OS` the right spots...

**Reply by FezMLG** [reply]
any updates on this topic?

**Reply by mikehardy** [reply]
No @FezMLG I don't see any.

**Reply by douglowder** [reply]
@mikehardy I'm the maintainer of the React Native repo for Apple TV and Android TV. I'd be happy to help with going through the code and at least getting a handle on the scope of the problem. It looks like Firebase podspecs do already support tvOS so I'm slightly optimistic that this could be done 😄

**Reply by mikehardy** [reply]
It should be nearly trivial actually, I'm thinking a small batch of ifdefs and done, and I'm happy to carry them in main. I think half or more of the effort (if we choose to do it) would be putting at least a compile check in a GitHub workflow, and ideally an e2e run of our test app (though I'm unsure of scope on that vs just running a build)

**Reply by douglowder** [reply]
Your E2E tests use Detox, which sadly does not work on tvOS. That said, I agree we can at least have a workflow that checks for a successful build of tvOS.

**Reply by mikehardy** [reply]
That alone would be a very positive step and I think cover 99% of potential future failures.

Separately - it appears that Detox may be made to work on tvOS as the previous was that underlying framework EarlyGrey did not work on tvOS, but then that changed and subsequently that framework was dropped anyway such that an in-repo DetoxSync framework is used.

As such, getting Detox to work on tvOS may be a possibliity now (and...just maybe...another one of those ifdef'ing festivals that is somehow easier then expected - could happen!).

Or it could be that on-device testing is better done in some other framework, preferably one that does not rely on the JSI (our current Detox/Jet system uses JSI...) and preferably one that works on all platforms (where I define that as android/ios/web/macOS/catalyst/windows - everything...). Not sure what that other platform would be but maybe some sort of webdriver appium thing - I only mention this because perhaps you have some domain knowledge here and could share wisdom, doesn't hurt to ask :-)

But concretely: just a build check? Still very useful + totally acceptable

**Reply by grigdodon** [reply]
any update on this?
when trying to build, I get:
```
Framework not found FirebaseAnalytics
Linker command failed with exit code 1 (use -v to see invocation)
```

**Reply by mikehardy** [reply]
Hey @grigdodon 👋 - nope, no concrete steps though I'm still open to any reasonable PR to get tvOS working, I continue to think it would just be a matter of ifdefing things out and then a build check that copies our current ios e2e yml and does everything except the actual e2e run.

When I see errors like yours though I get worried - I wonder if that's related to using unsupported `modular_headers` Podfile workarounds that will never work for react-native-firebase, instead of using `use_frameworks! :linkage => :static` ? If not that, then perhaps analytics isn't supported for tvOS, I haven't looked at the specific list of modules that work or don't work in a while

**Reply by dev-folks** [reply]
Hey!
I also spent a lot of time trying to find a successful case of using firebase for react-native-tvos. It seems that the user request for this is negligible - too few questions and information...

https://github.com/react-native-tvos/react-native-tvos/issues/811

**Reply by mikehardy** [reply]
I just took a look at this myself as I'd like to see react-native-tvos supported (and react-native-web, supported well for that matter vs just as shim-hacks like I've done)

I do not currently have a way to test tvOS support (though I just did another look at the 2 open feature requests in react-native-test-app and tried to confirm some assumptions there about adding it).

If you have a way to test current react-native-tvos or are willing to spend the time to make a skeleton repository (that I could just drop a google services info plist/json set into) then we could get it building I'm sure. In other repositories I manage, it's just a matter of doing the right compiler ifdefs to exclude API usage that is invalid on tvOS platform

**Comment by douglowder** [comment]
I'll take a look to see if the underlying Firebase Cocoapods dependencies actually support tvOS. If so, then we can probably do this.

**Reply by adamkoch** [reply]
I've been using `react-native-firebase` with `react-native-tvos` for a little while. I use `patch-package` to do 2 things:
 - update the podspecs to support tvos
 - update `firebase-auth` specifically to skip over an unsupported APIs in tvOS (you may need to double check this for your use-case because I blanket removed a bunch of stuff)

Here are the patches in case anyone wants to give it a go. They work fine with latest version `21.3.0` even though they were written for prior version.

[@react-native-firebase+storage+20.3.0.patch](https://github.com/user-attachments/files/17603937/%40react-native-firebase%2Bstorage%2B20.3.0.patch)
[@react-native-firebase+functions+20.3.0.patch](https://github.com/user-attachments/files/17603938/%40react-native-firebase%2Bfunctions%2B20.3.0.patch)
[@react-native-firebase+firestore+20.3.0.patch](https://github.com/user-attachments/files/17603939/%40react-native-firebase%2Bfirestore%2B20.3.0.patch)
[@react-native-firebase+auth+20.3.0.patch](https://github.com/user-attachments/files/17603940/%40react-native-firebase%2Bauth%2B20.3.0.patch)
[@react-native-firebase+app+21.3.0.patch](https://github.com/user-attachments/files/17603941/%40react-native-firebase%2Bapp%2B21.3.0.patch)
[@react-native-firebase+app-check+20.3.0.patch](https://github.com/user-attachments/files/17603943/%40react-native-firebase%2Bapp-check%2B20.3.0.patch)

**Reply by douglowder** [reply]
Thanks @adamkoch ! I took a quick look today, and created a brand new Expo TV app with SDK 51 and was able to add app and analytics packages with no issues. tvOS seems to compile correctly.

I had to add some of the Google and Firebase pods as extra pods in the Expo app config, as they require modular_headers=true.

See the commits in https://github.com/douglowder/FirebaseTV .

**Reply by mikehardy** [reply]
I'd be happy to get these merged in for you @adamkoch if you posted PRs?

**Reply by mikehardy** [reply]
@douglowder "they require modular_headers=true" uh-oh - we usually discourage that here in the form of "you are not supported if you do that", will have to investigate. Should be `use_frameworks! :linkage => :static`, not modular_headers anywhere...

**Reply by douglowder** [reply]
LOL I'm just doing what Cocoapods told me to do 😀 

I'll see if I can change that in the Expo config.

**Reply by mikehardy** [reply]
@douglowder you can, and I think this may be where react-native-tvos has some small problems, however, that is/was an old issue in your repo, was rooted in an upstream issue, and is I think all fixed now...

https://docs.expo.dev/versions/latest/sdk/build-properties/#pluginconfigtypeios

**Reply by adamkoch** [reply]
> I had to add some of the Google and Firebase pods as extra pods in the Expo app config, as they require modular_headers=true.

I don't have modular_headers=true but I did set useFrameworks static and also include the plugins from app, auth, app-check  `app.config.ts` (as well as specify firebase plist and json files as normal):
```
    '@react-native-firebase/app',
    '@react-native-firebase/auth',
    '@react-native-firebase/app-check',
    [
      'expo-build-properties',
      {
        ios: {
          useFrameworks: 'static',
        },
      },
    ],
```

**Reply by adamkoch** [reply]
> I'd be happy to get these merged in for you @adamkoch if you posted PRs?

I think the podspec modifications are straightforward, happy to put PRs up for those. The one that i'm less sure about is the auth modifications to `RNFBAuthModule.m` as I have zero experience with objective-c and basically just hacked through it to get it working :)

**Reply by mikehardy** [reply]
Just post up what you've got - it can't be too crazy. I've done tvOS in a few repos now, and I should be able to figure something out, but knowing I'm starting with a base that's working (for you) is an excellent first step given testing is difficult

**Reply by adamkoch** [reply]
Sure, will do shortly 👍

**Reply by adamkoch** [reply]
https://github.com/invertase/react-native-firebase/pull/8109

**Reply by douglowder** [reply]
Thanks @adamkoch I'll have a look

**Reply by mikehardy** [reply]
#8111 has what I think is final merge-worthy state on this - I just extended the style you posted, Adam, to all supported firebase components, and switched the auth ifdef compile guards to ios-only to match upstream - pretty simple really, happy for such a small change to enable an entire platform

**Reply by adamkoch** [reply]
Amazing, thank you!

**Reply by dev-folks** [reply]
Oh yes, that's amazing, you are legends!

**Comment by aleksandrsakov** [comment]
Hello guys, as I see you merged support for tvOS, but i can not manage to build project with React Native CLI 0.77 and RNFB v. 22.2.1.
My package.json
```
    "@react-native-firebase/app": "^22.2.1",
    "@react-native-firebase/firestore": "^22.2.1",
    "react-native": "npm:react-native-tvos@0.77.0-0",
```
I added in Podfile
```
use_frameworks! :linkage => :static
$RNFirebaseAsStaticFramework = true
```
And import Firebase and FirebaseApp.configure() in AppDelegate.swift

pod install --repo-update works fine, but after that when I am running buid ios, i got a tone of errors like on a screen
<img width="937" alt="image" src="https://github.com/user-attachments/assets/509ccf4a-48d4-4598-bc38-675fbf0c5498" />

Here is my full Podfile
```
# Resolve react_native_pods.rb with node to allow for hoisting
require Pod::Executable.execute_command('node', ['-p',
  'require.resolve(
    "react-native/scripts/react_native_pods.rb",
    {paths: [process.argv[1]]},
  )', __dir__]).strip

platform :tvos, min_ios_version_supported
prepare_react_native_project!

linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end

target 'nugs' do
  config = use_native_modules!
  use_frameworks! :linkage => :static
  $RNFirebaseAsStaticFramework = true

  use_react_native!(
    :path => config[:reactNativePath],
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  post_install do |installer|
    # https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/react_native_pods.rb#L197-L202
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false,
      # :ccache_enabled => true
    )
  end
end
```

Maybe you know what i missing? Can you help me here please? @mikehardy  @adamkoch  @douglowder

**Comment by adamkoch** [comment]
Not sure if this will be much help, but I find our set up to be quite fragile and breaks easily. Once it's working I don't touch it unless I absolutely have to.

My current set up uses:
```
"react-native": "npm:react-native-tvos@~0.76.9-0",
"@react-native-firebase/app": "^22.1.0",
```

I am NOT on the RN new architecture. I did try to upgrade to it in the past but it broke things pretty badly. Granted I didn't spend that long working on it.

I also use Expo and its managed workflow which generates the ios directory on the fly so I'm not configuring my podfile myself.
